added ReduceTrait and FlattenSequence

lukas-folle-snkeos opened the pull request:
Fixes #8528.

### Description

This PR adds the `FlatttenSequence` transform (a flavor of the also newly added `ReduceTrait`) which can flatten a nested data structure by one level. This way, #8528 can be tackled without the need to change the `apply_transform` of `Compose` significantly.

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

lukas-folle-snkeos committed:
DCO Remediation Commit for Lukas Folle <lukas.folle@snke.com>

I, Lukas Folle <lukas.folle@snke.com>, hereby add my Signed-off-by to this commit: e0cda55d4ad86efc74e912c023d9bcc5f6d12608

Signed-off-by: Lukas Folle <lukas.folle@snke.com>

lukas-folle-snkeos committed:
added list extend to MultiSampleTrait

lukas-folle-snkeos committed:
Merge branch 'dev' of github.com:lukas-folle-snkeos/MONAI into dev

lukas-folle-snkeos committed:
DCO Remediation Commit for Lukas Folle <lukas.folle@snke.com>

I, Lukas Folle <lukas.folle@snke.com>, hereby add my Signed-off-by to this commit: e0cda55d4ad86efc74e912c023d9bcc5f6d12608

Signed-off-by: Lukas Folle <lukas.folle@snke.com>

lukas-folle-snkeos committed:
fixed type errors

lukas-folle-snkeos committed:
Merge branch 'dev' of github.com:lukas-folle-snkeos/MONAI into dev

lukas-folle-snkeos committed:
DCO Remediation Commit for Lukas Folle <lukas.folle@snke.com>

I, Lukas Folle <lukas.folle@snke.com>, hereby add my Signed-off-by to this commit: e0cda55d4ad86efc74e912c023d9bcc5f6d12608

Signed-off-by: Lukas Folle <lukas.folle@snke.com>

lukas-folle-snkeos committed:
Merge branch 'dev' of github.com:lukas-folle-snkeos/MONAI into dev

lukas-folle-snkeos committed:
DCO Remediation Commit for Lukas Folle <lukas.folle@snke.com>

I, Lukas Folle <lukas.folle@snke.com>, hereby add my Signed-off-by to this commit: e0cda55d4ad86efc74e912c023d9bcc5f6d12608

Signed-off-by: Lukas Folle <lukas.folle@snke.com>

lukas-folle-snkeos committed:
Merge branch 'dev' of github.com:lukas-folle-snkeos/MONAI into dev

DCO Remediation Commit for Lukas Folle <lukas.folle@snke.com>

I, Lukas Folle <lukas.folle@snke.com>, hereby add my Signed-off-by to this commit: eeb7e12d604a4c46f5b44d7484b40aa9cac6e9d3

Signed-off-by: Lukas Folle <lukas.folle@snke.com>

lukas-folle-snkeos committed:
DCO Remediation Commit for Lukas Folle <lukas.folle@snke.com>

I, Lukas Folle <lukas.folle@snke.com>, hereby add my Signed-off-by to this commit: e0cda55d4ad86efc74e912c023d9bcc5f6d12608

Signed-off-by: Lukas Folle <lukas.folle@snke.com>

lukas-folle-snkeos committed:
avoided breaking map_item functionality

DCO Remediation Commit for Lukas Folle <lukas.folle@snke.com>

I, Lukas Folle <lukas.folle@snke.com>, hereby add my Signed-off-by to this commit: e0cda55d4ad86efc74e912c023d9bcc5f6d12608

Signed-off-by: Lukas Folle <lukas.folle@snke.com>

lukas-folle-snkeos committed:
fixed wrong type annotation

DCO Remediation Commit for Lukas Folle <lukas.folle@snke.com>

I, Lukas Folle <lukas.folle@snke.com>, hereby add my Signed-off-by to this commit: e0cda55d4ad86efc74e912c023d9bcc5f6d12608

Signed-off-by: Lukas Folle <lukas.folle@snke.com>

lukas-folle-snkeos commented:
All remaining failing CI runs are not coming from this change.

ericspod committed:
Merge branch 'dev' into dev

ericspod commented on monai/transforms/transform.py:
```diff
@@ -143,10 +143,19 @@ def apply_transform(
+                # Only extend if we're at the leaf level (map_items_ == 1) and the transform
+                # actually returned a list (not preserving nested structure)
+                if isinstance(res_item, list) and map_items_ == 1:
+                    if not isinstance(item, (list, tuple)):
```
Within the containing block `res_item` will always be a list (line 151) and so this if statement's condition is always false. I'm not sure the logic is what you're expecting it to be here?

lukas-folle-snkeos replied:
@ericspod I am not sure I understand what you mean. In case `res_item` is `list`, `item` can still be something else.

ericspod replied:
Sorry for the delay, I might have not read correctly what was going on here. I think this block 151-157 can be simplified to:

```python
if isinstance(res_item, list) and map_items_ == 1 and not isinstance(item, (list, tuple)):
    res.extend(res_item)
else:
    res.append(res_item)
```

The result list is extended when `transform` turns a non-list/tuple `item` into a list specifically, and appended otherwise. I think what I'm not sure about is under what situation `item` would be a list/tuple, and even if it was why we'd append in that case. Am I missing something with that?

We should have an added test as well that would catch the case described in the original issue, ideally using existing transforms rather than dummy ones for testing. Thanks!

ericspod commented:
Hi @lukas-folle-snkeos thanks for the contribution! Please check the logic I flagged and we do need tests to check that this behaviour is correct. You can take the mentioned issue as an example of what to check for but I think we need to test this logic thoroughly and more clearly document what's going on in comments.

atbenmurray commented:
Taking a look.

lukas-folle-snkeos committed:
added test for many multisample transforms; refactored code

lukas-folle-snkeos committed:
DCO Remediation Commit for Lukas Folle <lukas.folle@snke.com>

I, Lukas Folle <lukas.folle@snke.com>, hereby add my Signed-off-by to this commit: 2d5877455022444675e200afee36077a709fa784

Signed-off-by: Lukas Folle <lukas.folle@snke.com>

lukas-folle-snkeos commented:
Thanks for the review @ericspod! I added a test with a real transform and refactored the code as suggested.

lukas-folle-snkeos committed:
Merge branch 'dev' of github.com:lukas-folle-snkeos/MONAI into dev

lukas-folle-snkeos committed:
DCO Remediation Commit for Lukas Folle <lukas.folle@snke.com>

I, Lukas Folle <lukas.folle@snke.com>, hereby add my Signed-off-by to this commit: 2d5877455022444675e200afee36077a709fa784

Signed-off-by: Lukas Folle <lukas.folle@snke.com>

lukas-folle-snkeos committed:
DCO Remediation Commit for Lukas Folle <lukas.folle@snke.com>

I, Lukas Folle <lukas.folle@snke.com>, hereby add my Signed-off-by to this commit: 2d5877455022444675e200afee36077a709fa784

Signed-off-by: Lukas Folle <lukas.folle@snke.com>

Merge branch 'dev' of github.com:lukas-folle-snkeos/MONAI into dev

lukas-folle-snkeos committed:
added slight cleanup and additional test

DCO Remediation Commit for Lukas Folle <lukas.folle@snke.com>

I, Lukas Folle <lukas.folle@snke.com>, hereby add my Signed-off-by to this commit: 2d5877455022444675e200afee36077a709fa784

Signed-off-by: Lukas Folle <lukas.folle@snke.com>

atbenmurray commented:
Hi folks. I've taken an initial look at this. I'm always a little afraid of these kind of changes that mandate a particular behaviour. Are there no situations where someone might multi-sample something and then multi-sample the multi-sampled thing and expect the lists to be nested? Should this be fixed in execute_compose instead? Please give me a couple of hours today to take a look at that.

lukas-folle-snkeos committed:
Merge branch 'dev' into dev

lukas-folle-snkeos commented:
@atbenmurray @ericspod could you please check the status of this PR again?

atbenmurray commented:
@lukas-folle-snkeos My apologies; taking a look now

atbenmurray commented:
Okay, so apologies for the long wait. I finally managed to get my head into it properly:

I'd like to suggest a slight change. I think that, if a user wants to run multisamples on top of multisamples there will be scenarios in which they don't want to next the output and scenarios in which they do. I'd rather we didn't remove the ability to do so:

I propose that we do the following instead:
1. Add ReduceTrait to the list of traits
2. Tweak the code in apply_transform, but slightly differently
3. Add a FlattenSequence transform that can be inserted when the user wants multiply nested transforms to not generate nested lists

```
        # modified apply_transform snippet that is ReduceTrait aware
        map_items_ = int(map_items) if isinstance(map_items, bool) else map_items
        if isinstance(data, (list, tuple)) and map_items_ > 0 and not isinstance(transform, ReduceTrait):
            return [
                apply_transform(transform, item, map_items_ - 1, unpack_items, log_stats, lazy, overrides)
                for item in data
            ]
        return _apply_transform(transform, data, unpack_items, lazy, overrides, log_stats)
```

```
class ReduceTrait:
    """
    An interface to indicate that the transform has the capability to reduce multiple samples
    into a single sample.
    This interface can be extended from by people adapting transforms to the MONAI framework as well
    as by implementors of MONAI transforms.
    """

    pass
```

```
class FlattenSequence(mt.Transform, mtt.ReduceTrait):
    def __init__(self):
        super().__init__()

    def __call__(self, data):
        if isinstance(data, (list, tuple)):
            if len(data) == 0:
                return data
            if isinstance(data[0], (list, tuple)):
                return [item for sublist in data for item in sublist]
        return data
```

The footprint has a slightly larger change but then we don't lose the flexibility

It would be used as follows:

```    transforms_2 = [
        mt.CenterSpatialCrop((128, 128)),
        mt.RandSpatialCropSamples((96, 96), num_samples=3),
        mt.RandSpatialCropSamples((64, 64), num_samples=2),
        FlattenSequence(),
 ```
 
 We can make interesting uses of ReduceTrait, allowing it to be a trait that can do things like aggregations, catting, etc, as well as flattening lists. WDYT?

lukas-folle-snkeos committed:
Merge branch 'dev' into dev

lukas-folle-snkeos commented:
@atbenmurray I like that idea; looks a whole lot cleaner to me compared to the initially proposed change! I will bring the changes into this PR.

lukas-folle-snkeos committed:
changed compose to explicit flattening

Signed-off-by: Lukas Folle <lukas.folle@snke.com>

lukas-folle-snkeos committed:
added documentation

Signed-off-by: Lukas Folle <lukas.folle@snke.com>

lukas-folle-snkeos committed:
fixed doc build; fixed isort

Signed-off-by: Lukas Folle <lukas.folle@snke.com>

lukas-folle-snkeos committed:
added type hints and fixed potential bug

Signed-off-by: Lukas Folle <lukas.folle@snke.com>

lukas-folle-snkeos committed:
formatted

Signed-off-by: Lukas Folle <lukas.folle@snke.com>

lukas-folle-snkeos committed:
ignored mypy error

Signed-off-by: Lukas Folle <lukas.folle@snke.com>

lukas-folle-snkeos commented:
@atbenmurray do you want to have a look at the proposed changes?

atbenmurray reviewed (approved):
Looks good to me

lukas-folle-snkeos commented:
@atbenmurray not sure how to make the blossom-ci run.. Seems to be required to merge.

atbenmurray commented:
> @atbenmurray not sure how to make the blossom-ci run.. Seems to be required to merge.

I don't have the powers, sadly. @ericspod @KumoLiu @Nic-Ma?

lukas-folle-snkeos commented:
bump @ericspod @KumoLiu @Nic-Ma @atbenmurray

ericspod commented:
@KumoLiu I think we're good to go here, please trigger blossom when you can, thanks.

KumoLiu committed:
Merge branch 'dev' into dev

KumoLiu commented:
/build

lukas-folle-snkeos committed:
Merge branch 'dev' into dev

KumoLiu commented:
/build
